src/tpm2_swtpm.c

```diff
@@ -0,0 +1,289 @@
+        fd = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
+        if (fd == -1)
+            continue;
+
+        if (connect(fd, rp->ai_addr, rp->ai_addrlen) == -1) {
+            close(fd);
+        } else {
+            break;
+        }
+    }
+    freeaddrinfo(result);
+
+    if (rp != NULL) {
+        ctx->tcpCtx.fd = fd;
+        rc = TPM_RC_SUCCESS;
+    } else {
+      #ifdef DEBUG_WOLFTPM
+        printf("Failed to connect to %s %s\n", host, port);
+      #endif
+    }
+
+    return rc;
+}
+
+static TPM_RC SwTpmDisconnect(TPM2_CTX* ctx)
+{
+    TPM_RC rc = TPM_RC_SUCCESS;
+    uint32_t tss_cmd;
+
+    if (ctx == NULL || ctx->tcpCtx.fd <= 0) {
+        return BAD_FUNC_ARG;
+    }
+
+    /* end swtpm session */
+    tss_cmd = htonl(TPM_SESSION_END);
+    rc = SwTpmTransmit(ctx, &tss_cmd, sizeof(uint32_t));
+    #ifdef WOLFTPM_DEBUG_VERBOSE
+    if (rc != TPM_RC_SUCCESS) {
+        printf("Failed to transmit SESSION_END\n");
+    }
+    #endif
+
+    if (0 != close(ctx->tcpCtx.fd)) {
+        rc = SOCKET_ERROR_E;
+
+        #ifdef WOLFTPM_DEBUG_VERBOSE
+        printf("Failed to close fd %d, got errno %d ="
+               "%s\n", ctx->tcpCtx.fd, errno, strerror(errno));
+        #endif
+    }
+
+    ctx->tcpCtx.fd = -1;
+
+    return rc;
+}
+
+/* Talk to a TPM through socket
+ * return TPM_RC_SUCCESS on success,
+ *        SOCKET_ERROR_E on socket errors,
+ *        TPM_RC_FAILURE on other errors
+ */
+int TPM2_SWTPM_SendCommand(TPM2_CTX* ctx, TPM2_Packet* packet)
+{
+    int rc = TPM_RC_FAILURE;
+    int rspSz = 0;
+    uint32_t tss_word;
+
+    if (ctx == NULL) {
+        return BAD_FUNC_ARG;
+    }
+
+    if (ctx->tcpCtx.fd <= 0) {
+        rc = SwTpmConnect(ctx, TPM2_SWTPM_HOST, TPM2_SWTPM_PORT);
+    }
+
+#ifdef WOLFTPM_DEBUG_VERBOSE
+    printf("Command size: %d\n", packet->pos);
+    TPM2_PrintBin(packet->buf, packet->pos);
+#endif
+
+    /* send start */
+    tss_word = htonl(TPM_SEND_COMMAND);
+    if (rc == TPM_RC_SUCCESS) {
+        rc = SwTpmTransmit(ctx, &tss_word, sizeof(uint32_t));
+    }
+
+    /* locality */
+    if (rc == TPM_RC_SUCCESS) {
+        rc = SwTpmTransmit(ctx, &ctx->locality, sizeof(uint8_t));
+    }
+
+    /* buffer size */
+    tss_word = htonl(packet->pos);
+    if (rc == TPM_RC_SUCCESS) {
+        rc = SwTpmTransmit(ctx, &tss_word, sizeof(uint32_t));
+    }
+
+    /* Send the TPM command buffer */
+    if (rc == TPM_RC_SUCCESS) {
+        rc = SwTpmTransmit(ctx, packet->buf, packet->pos);
+    }
+
+    /* receive response */
+    if (rc == TPM_RC_SUCCESS) {
+        rc = SwTpmReceive(ctx, &tss_word, sizeof(uint32_t));
+        rspSz = ntohl(tss_word);
+        if (rspSz > packet->size) {
+            #ifdef WOLFTPM_DEBUG_VERBOSE
+            printf("Response size(%d) larger than command buffer(%d)\n",
+                   rspSz, packet->pos);
+            #endif
+            rc = SOCKET_ERROR_E;
+        }
+    }
+
+    /* TODO: could hang as currently implemented, but is not TSS compliant */
```

Comment:
Also is this something that should be fixed or maybe the comment expanded to explain why? I think its because it could block there without timeout, right?

Reply:
Hope the new comment helps.